ui/src/components/FileUpload.tsx

```diff
@@ -0,0 +1,107 @@
+import React, { useCallback, ChangeEvent, useEffect } from 'react';
+import { PutObjectCommand } from '@aws-sdk/client-s3';
+import { dateToDa, deSig } from '@urbit/api';
+import { useFileStore, useStorage } from '@/state/storage';
+import { Upload, UploadInputProps } from '@/types/storage';
+import api from '../api';
+
+const UploadInput = React.forwardRef<HTMLInputElement, UploadInputProps>(
+  (props, ref) => {
+    const { multiple, ...rest } = props;
+    const { files, createClient, ...fs } = useFileStore();
+    const { s3, ...storage } = useStorage();
+    const { credentials } = s3;
+
+    useEffect(() => {
+      async function init() {
+        useStorage.getState().initialize(api);
+      }
+      init();
+    }, []);
+
+    useEffect(() => {
+      const hasCredentials =
+        credentials?.accessKeyId &&
+        credentials?.endpoint &&
+        credentials?.secretAccessKey;
+      if (hasCredentials) {
+        const connect = async () => {
+          const client = createClient(credentials);
+          return client;
+        };
+        connect()
+          .then(() => fs.setStatus('success'))
+          .catch((error: unknown) => {
+            console.error(error);
+          });
+      }
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [createClient, credentials]);
+
+    const uploadFile = useCallback(
+      async (upload: Upload, idx: number) => {
+        const timestamp = deSig(dateToDa(new Date()));
+        const { file } = upload;
+        if (fs.client) {
+          fs.setFileStatus([idx, 'loading']);
+          const uploadData = await fs.client
+            .send(
+              new PutObjectCommand({
+                Bucket: s3.configuration.currentBucket,
+                Key: `${window.ship}/${timestamp}-${file.name}`,
+                Body: file,
+                ContentType: file.type,
+                ContentLength: file.size,
+                ACL: 'public-read',
+              })
+            )
+            .then(() => {
```

Comment:
please don't mix `await` and `.then()` unless absolutely necessary

Reply:
Cleaned this up, removed the `await`